Playwright: Use selector instead to locate the `close tour` button.

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This PR uses the new Locator API in order to click the `close tour` button, which is causing many issues with the editor related tests.

Key changes:
- replace the `frame.click` call with `locator.click` call in `dismissWelcomeTourIfPresent`.
- use Locator API to locate the iframed editor in `getEditorFrame`.

#### Testing instructions

In my local test, it was able to locate the `close tour` button where the previous approach failed to do so.

Related to #57460

worldomonation committed:
Use selector instead to locate the `close tour` button.

worldomonation committed:
Use locator for editor frame.

worldomonation commented:
@WunderBart What do you think of this approach instead of https://github.com/Automattic/wp-calypso/pull/58141?

WunderBart commented:
> @WunderBart What do you think of this approach instead of #58141?
> 
> Note, I plan to integrate some of your suggestions left in the other PR, such as [#58141 (comment)](https://github.com/Automattic/wp-calypso/pull/58141#discussion_r760290643).

I think there might be another way! I haven't tested it yet via the tests, but the Welcome Tour state can be controlled via the app state. In the console, we can toggle it with:

```js
wp.data.dispatch( 'automattic/wpcom-welcome-guide' ).setShowWelcomeGuide( false ); 
```

It will send a POST request and turn the A8c's Welcome Tour off. Maybe it would be enough to just evaluate it e.g. in a `beforeAll` hook?

worldomonation committed:
Use `page.on` to intercept and inspect network responses, and if a `nux?envelope=1` request is found call additional actions depending on the outcome of the value present in the response.

worldomonation commented:
> > @WunderBart What do you think of this approach instead of #58141?
> > Note, I plan to integrate some of your suggestions left in the other PR, such as [#58141 (comment)](https://github.com/Automattic/wp-calypso/pull/58141#discussion_r760290643).
> 
> I think there might be another way! I haven't tested it yet via the tests, but the Welcome Tour state can be controlled via the app state. In the console, we can toggle it with:
> 
> ```js
> wp.data.dispatch( 'automattic/wpcom-welcome-guide' ).setShowWelcomeGuide( false ); 
> ```
> 
> It will send the nux POST request and turn the A8c's Welcome Tour off. Maybe it would be enough to just evaluate it e.g. in a `beforeAll` hook?

That's an interesting option and will keep it in mind in case my approach I just pushed in 8b44cd61cc942cf452cb00d940d4739941cbbfa0 fails to work!

worldomonation commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -74,14 +74,28 @@ export class GutenbergEditorPage {
+		// `nux?_envelope=1` request has been completed.
+		// On the other hand, `page.on` is able to monitor every request, including the
+		// one to the `nux` endpoint.
+		this.page.on( 'response', async ( response ) => {
+			if ( ! response.url().includes( 'nux?_envelope=1' ) ) {
+				return;
+			}
+
+			const body: { [ index: string ]: never } = await response.json();
+			if ( body[ 'body' ][ 'show_welcome_guide' ] === true ) {
+				await this.dismissWelcomeTourIfPresent();
+			}
```
As explained in the code comments, this construct intercepts and inspects every response coming in to determine whether any of them are sent from the `nux` endpoint.

I've tried to implement this originally with `waitForResponse`, but as noted in the comments it would hang until the timeout was reached. After debugging for a bit it appears that `waitForResponse` either starts monitoring too late in the process, or is not able to properly intercept the `nux` request.

Whichever the case may be, I inadvertently verified that `waitForResponse` cannot be used here when I:

- launched a debugging session with `waitForResponse`
- it hung, as expected
- reload the window
- `waitForRequest` promise completed and execution continued

To deal with the above, one option is to wrap every call to `GutenbergEditorPage.waitUntilLoaded()` with a `Promise.all` containing `waitForRequest`, but that would be leaking out matters internal to GutenbergEditorPage outside of the class so that's a no go.

Which is why I ended up using `page.on`, and it works really well.

worldomonation replied:
Removal of a catch-all `try/catch` block with forced 5 second wait also ended up reducing execution time.

worldomonation replied:
Looks like this approach failed with `pecs/specs-playwright/wp-blocks__media-spec.ts`. Investigating.

worldomonation committed:
Use 'requestfinished' instead of 'response' to wait for the body to be fully downloaded.

WunderBart commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -74,14 +74,33 @@ export class GutenbergEditorPage {
+				return;
+			}
+
+			const body: { [ index: string ]: never } = await response.json();
+			if ( body[ 'body' ][ 'show_welcome_guide' ] === true ) {
+				await this.dismissWelcomeTourIfPresent();
+			}
```
Let's use optional chaining here just to be extra safe:

```suggestion
			interface NuxPayload {
				body: {
					show_welcome_guide: boolean;
				};
			}
			const payload = ( await response.json() ) as NuxPayload;
			if ( payload?.body?.show_welcome_guide === true ) {
				await this.dismissWelcomeTour();
			}
```

worldomonation replied:
Interesting, I'll try this out.

worldomonation replied:
045a0ef365865b62ae5827960ff80cc17625aad0

WunderBart commented on packages/calypso-e2e/src/lib/pages/gutenberg-editor-page.ts:
```diff
@@ -90,11 +109,9 @@ export class GutenbergEditorPage {
 	 */
 	async dismissWelcomeTourIfPresent(): Promise< void > {
```
Looks like we can remove the `IfPresent` bit here?

```suggestion
	async dismissWelcomeTour(): Promise< void > {
```

worldomonation replied:
Totally, a welcome change!

worldomonation replied:
Fixed in 045a0ef365865b62ae5827960ff80cc17625aad0.

WunderBart reviewed (approved):
Looks promising with the `nux` request handling! 🤞 

Just 2 small suggestions from me 💪

worldomonation committed:
Incorporate feedback from PR:
- rename `dismissWelcomeTourIfPresent`.
- use interface within the `requestfinished` body.

worldomonation commented:
Merging this as all E2E tests (incl. Quarantine and Pre-Release) are passing.
